doc: ipc: Add detailed protocol specs for ICMsg backend

doki-nordic opened the pull request:
The details of how the ICMsg communication is done, are hidden inside the source code. This specification adds details about the ICBMsg protocol.

doki-nordic commented:
Direct link to docs:
https://builds.zephyrproject.io/zephyr/pr/78045/docs/services/ipc/ipc_service/backends/ipc_service_icmsg.html#detailed-protocol-specification

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+Shared Memory Region Organization
+---------------------------------
+
+If data caching is enabled, the shared memory region provided to ICMsg must be
+aligned according to the cache requirement.
```
```suggestion
If data caching is enabled, the shared memory region provided to ICMsg must be aligned according to the cache requirement.
```

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+
+The shared memory region is entirely used by a single FIFO.
+It contains read and write indexes followed by the data buffer.
+Detailed structure is contained in the following table.
```
```suggestion
Detailed structure is contained in the following table:
```

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+     - n/a
+     - Circular buffer containing actual bytes to transfer.
+
+This is usual FIFO with a circular buffer, so:
+
+* indexes (``rd_idx`` and ``wr_idx``) are wrapped around when they reach the end of the ``data`` buffer,
+* when ``rd_idx == wr_idx``, the FIFO is empty,
+* it has one byte less capacity than the ``data`` buffer length.
```
```suggestion
This is usual FIFO with a circular buffer, so indexes (``rd_idx`` and ``wr_idx``) are wrapped around under the following conditions:

* When they reach the end of the ``data`` buffer.
* When ``rd_idx == wr_idx``, the FIFO is empty,
* When it has one byte less capacity than the ``data`` buffer length.
```

Please check if the meaning has changed here.

doki-nordic replied:
It is different meaning. I was trying to say that this is usual FIFO, but such statement is not strict enough, so I added those 3 points characterizing it for clarification. Each of them is independent, and you joined them which is wrong. Maybe rewrite the points as follows?
```
This is usual FIFO with a circular buffer:

* The indexes (``rd_idx`` and ``wr_idx``) are wrapped around when they reach the end of the ``data`` buffer.
* The FIFO is empty if ``rd_idx == wr_idx``.
* The FIFO has one byte less capacity than the ``data`` buffer length.
```

divipillai replied:
Agree.

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+Packets are sent over the FIFO described in the above section.
+One packet can be wrapped around if it occurs at the end of the FIFO buffer.
+
+The packet structure is following.
```
```suggestion
The following is the packet structure:
```

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+
+The packet send procedure is following:
+
+#. Check if the packet will fit into the buffer.
```
```suggestion
#. Check if the packet fits into the buffer.
```

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+     - n/a
+     - Padding is added to align the total packet size to 4 bytes.
+
+The packet send procedure is following:
```
```suggestion
The packet send procedure is following:
```

doki-nordic replied:
I see no change there.

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+Initialization
+--------------
+
+The initialization sequence is following:
```
```suggestion
The initialization sequence is the following:
```

divipillai commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
+   It will arrive over the other pair (shared memory region and MBOX).
+#. Stop repeating the MBOX notification.
+
+After this, the ICMsg is bounded and it is ready to transfer packets.
```
```suggestion
After this, the ICMsg is bounded, and it is ready to transfer packets.
```

doki-nordic commented on doc/services/ipc/ipc_service/backends/ipc_service_icmsg.rst:
```diff
@@ -69,18 +69,127 @@ Bonding
 
  - :zephyr:code-sample:`ipc-icmsg`
+
+Detailed Protocol Specification
```
I disagree. The "Detailed Protocol Specification" is for users that need to know low level details about the protocol (which is very unusual case). The configuration is for all user. They do not need to know how the shared memory is organized and when MBOXes are used. They just need to assign resources for them using configuration. Shared memory and MBOXes are HW or system specific things (not ICMsg specific), so user should be familiar with them before reading "Detailed Protocol Specification".

divipillai replied:
Agree, thanks for the explanation.

doki-nordic committed:
doc: ipc: Add detailed protocol specs for ICMsg backend

The details of how the ICMsg communication is done, are hidden
inside the source code. This specification adds details about
the ICBMsg protocol.

Signed-off-by: Dominik Kilian <Dominik.Kilian@nordicsemi.no>
